curp/src/server/storage/wal/segment.rs

```diff
@@ -0,0 +1,452 @@
+    }
+
+    /// Syncs the file of this segment
+    pub(super) async fn sync_all(&mut self) -> io::Result<()> {
+        self.file.sync_all().await?;
+        self.io_state.fsynced();
+
+        Ok(())
+    }
+
+    /// Updates the seal index
+    pub(super) fn update_seal_index(&mut self, index: LogIndex) {
+        self.seal_index = self.seal_index.max(index);
+    }
+
+    /// Get the size of the segment
+    pub(super) fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// Checks if the segment is full
+    pub(super) fn is_full(&self) -> bool {
+        self.size >= self.size_limit
+    }
+
+    /// Gets the id of this segment
+    pub(super) fn id(&self) -> u64 {
+        self.segment_id
+    }
+
+    /// Gets the base log index of this segment
+    pub(super) fn base_index(&self) -> u64 {
+        self.base_index
+    }
+
+    /// Checks if the segment is redundant
+    ///
+    /// The segment is redundant if the `seal_index` is less than the `base_index`
+    pub(super) fn is_redundant(&self) -> bool {
+        self.seal_index < self.base_index
+    }
+
+    /// Gets the io state of this segment
+    pub(super) fn io_state(&self) -> IOState {
+        self.io_state
+    }
+
+    /// Gets the file name of the WAL segment
+    pub(super) fn segment_name(segment_id: u64, log_index: u64) -> String {
+        format!("{segment_id:016x}-{log_index:016x}{WAL_FILE_EXT}")
+    }
+
+    #[allow(clippy::doc_markdown)] // False positive for ASCII graph
+    /// Generate the header
+    ///
+    /// The header layout:
+    ///
+    /// 0      1      2      3      4      5      6      7      8
+    /// |------+------+------+------+------+------+------+------|
+    /// | Magic                     | Reserved           | Vsn  |
+    /// |------+------+------+------+------+------+------+------|
+    /// | BaseIndex                                             |
+    /// |------+------+------+------+------+------+------+------|
+    /// | SegmentID                                             |
+    /// |------+------+------+------+------+------+------+------|
+    /// | Checksum (32bytes) ...                                |
+    /// |------+------+------+------+------+------+------+------|
+    fn gen_header(base_index: LogIndex, segment_id: u64) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend(WAL_MAGIC.to_le_bytes());
+        buf.extend(vec![0; 3]);
+        buf.push(WAL_VERSION);
+        buf.extend(base_index.to_le_bytes());
+        buf.extend(segment_id.to_le_bytes());
+        buf.extend(get_checksum(&buf));
+        buf
+    }
+
+    /// Parse the header from the given buffer
+    #[allow(
+        clippy::unwrap_used, // Unwraps are used to convert slice to const length and is safe
+        clippy::integer_arithmetic,// Arithmetics cannot overflow
+        clippy::indexing_slicing // Index slicings are checked
+    )]
+    fn parse_header(src: &[u8]) -> Result<(LogIndex, u64), WALError> {
```

Comment:
Please add an unit test to cover the `parse_header` method.

Reply:
done